### PR TITLE
Responder AI (with wandering patrols after initial search)

### PR DESCRIPTION
**Pull request by fnnbrr**
Had to restructure the class hierarchy for AI with this one, which I'll continue to do once it's time for friendly NPCs.

Mess around with it and see if anything breaks - @chadleytan, would you take a look at the HandleTargets method in Responder.cs?  The responder nurses don't properly pursue Birdie.

### CODE REVIEW

File: Assets/Scripts/NPCs/Vision.cs [hunk 1]
```diff
@@ -1,88 +1,88 @@
...
+
+	//	viewMesh.Clear();
+	//	viewMesh.vertices = vertices;
+	//	viewMesh.triangles = triangles;
+	//	viewMesh.RecalculateBounds();
 	//	viewMesh.RecalculateNormals();
-	//}
-}
+	//}
+}
```
Comment (chadleytan) [inline Assets/Scripts/NPCs/Vision.cs]: I think we can just delete this file. I started making this during game jam 2 but then put everything in FieldOfVision instead.

**Comment by chadleytan** [comment]
@jurcaua & @justineechen could you 2 have a look at this as well? I'll keep looking at it but I think you 2 are more acquainted with the chase/pursue functionality than me and it would be great to have more eyes on this too.

File: Assets/Scripts/Missions/MissionCafeteria1/MissionCafeteria1.cs [hunk 2]
```diff
@@ -108,7 +108,7 @@ private void SpawnEnemies(List<MissionEnemy> enemies)
             if (NavMesh.SamplePosition(enemy.spawnPosition, out NavMeshHit closestNavmeshHit, 10.0f, NavMesh.AllAreas))
             {
                 GameObject spawnedEnemy = Instantiate(enemy.prefab, closestNavmeshHit.position, Quaternion.Euler(enemy.spawnRotation));
-                NursePatrol patrol = Utils.GetRequiredComponent<NursePatrol>(spawnedEnemy, $"Enemy in MissionCafeteria1 does not have a NursePatrol component!");
+                Patroller patrol = Utils.GetRequiredComponent<Patroller>(spawnedEnemy, $"Enemy in MissionCafeteria1 does not have a NursePatrol component!");
```
Comment (jurcaua) [inline Assets/Scripts/Missions/MissionCafeteria1/MissionCafeteria1.cs]: Update the comment here too?

File: Assets/Scripts/NPCs/Chaser.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    [Header("Chaser")]
+    public bool chase = true;
+    public Transform targetTransform;
+    
+    [HideInInspector] public NavMeshAgent agent;
+
+    // Start is called before the first frame update
+    public void Start()
+    {
+        agent = GetComponent<NavMeshAgent>();
```
Comment (jurcaua) [inline Assets/Scripts/NPCs/Chaser.cs]: Can we use `Utils.GetRequiredComponent` just to be safe?

File: Assets/Scripts/NPCs/Patroller.cs [hunk 4]
```diff
@@ -76,30 +81,22 @@ void GotoNextPoint()
...
+                    if (WaypointPauseComplete())
+                    {
+                        GotoNextPoint();
+                    }
                     break;
-                case NurseStates.Chasing:
+                case ActionStates.Chasing:
                     ChaseTarget();
                     break;
-                default:
```
Comment (jurcaua) [inline Assets/Scripts/NPCs/Patroller.cs:101]: Switch statement should probably have a default statement right? Maybe we should also log something if this happens?

File: Assets/Scripts/NPCs/Patroller.cs [hunk 5]
```diff
@@ -48,20 +40,33 @@ public void SetPoints(List<Vector3> newPoints)
...
-        else if (numTargetsInRange == 0 && currentState != NurseStates.Patrolling)
+        else if (numTargetsInRange == 0 && currentState != ActionStates.Patrolling)
         {
             GotoNextPoint();
-            currentState = NurseStates.Patrolling;
+            currentState = ActionStates.Patrolling;
         }
     }
 
+    bool WaypointPauseComplete()
```
Comment (jurcaua) [inline Assets/Scripts/NPCs/Patroller.cs:58]: Not needed for this pull request, but I'd keep in mind future updates to include `EnemyWaypoint.stayTime` from `MissionCafeteria1.cs` to have this wait time changeable, and per-waypoint.

File: Assets/Scripts/NPCs/Responder.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        waypointPauseTimer = waypointPauseSec;
+        return true;
+    }
+    
+    void GotoNextPoint()
+    {
+        float wanderRadius = wanderBounds.radius;
+        Vector3 randomMovement = Random.insideUnitSphere * wanderRadius + wanderBounds.center;
+
+        NavMesh.SamplePosition(randomMovement, out NavMeshHit navHit, wanderRadius, -1);
```
Comment (jurcaua) [inline Assets/Scripts/NPCs/Responder.cs]: This function returns a bool for if it succeeded for not, can we check to make sure this worked or not? Never know when the wanderBounds is moved by accident and this ends up confusing everyone why wandering isn't working.

File: Assets/Scripts/NPCs/Responder.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                    if (WaypointPauseComplete())
+                    {
+                        GotoNextPoint();
+                    }
+                    break;
+                case ActionStates.Chasing:
+                    ChaseTarget();
+                    break;
+                case ActionStates.Responding:
+                    currentState = ActionStates.Wandering;
```
Comment (jurcaua) [inline Assets/Scripts/NPCs/Responder.cs:89]: I may be reading stuff incorrectly... but what is happening when we're wandering? From wandering I do not see logic to move it to another state, but I may be overlooking something?

File: Assets/Scripts/NPCs/Patroller.cs (hunk 4 above)
Comment (jurcaua) [reply Assets/Scripts/NPCs/Patroller.cs:101]: Throwing an exception can be a good idea like you did below.

**Review by jurcaua — CHANGES_REQUESTED** [review]
Some questions, but otherwise when I test it on my local machine it seems to work as expected! Nice stuff for sure.

Indeed the chase seems to be off. I'd ignore that for this pull request and then someone can look into after this is merged in. Let's create an issue for it.

**Comment by jurcaua** [comment]
Oh, just remembered seeing this when I tested, there were some places where the enemy view was being obstructed by some invisible object I believe? I don't have any screenshots but, I'd make sure the obstacle mask and stuff are set for all the enemies, and that objects in the scene that should be treated like obstacles are using the correct layer.

File: Assets/Scripts/Missions/MissionCafeteria1/MissionCafeteria1.cs (hunk 2 above)
Comment (fnnbrr) [reply Assets/Scripts/Missions/MissionCafeteria1/MissionCafeteria1.cs]: Oops yeah good catch, I just hit "refactor" without the "refactor string occurences" option on

File: Assets/Scripts/NPCs/Patroller.cs (hunk 4 above)
Comment (fnnbrr) [reply Assets/Scripts/NPCs/Patroller.cs:101]: Will do, I'll add a failsafe switch case that throws an exception.

**Comment by fnnbrr** [comment]
Yeah that was another issue, I had to change the vision cones to detect the "default" layers instead of the "obstacle" layer because everything in the scene was already assigned the "restricted" layer.  Unless something can be part of 2 layers at once, we'll have to turn one of them into a tag or something.

**Comment by jurcaua** [comment]
@FinnbarrOC I actually do not think we're using the Restricted / Free-roam layers at all, so I think those can be safely changed! So feel free to change everything to obstacle and default and stuff, and make the enemies view detect the obstacle layer

File: Assets/Scripts/NPCs/Responder.cs (hunk 7 above)
Comment (fnnbrr) [reply Assets/Scripts/NPCs/Responder.cs:89]: The code to handle Wandering is the same as the code used to handle Patrolling in Patroller.cs.
It just keeps picking new destinations indefinitely.

**Review by jurcaua — APPROVED** [review]
Everything looks good to me!! IGN 10/10 would review and leave small comments again

**Pull request merged**